<html><head>
    
    <title>Eh?: familyview</title>
  <h3 align="center">Ensembl FamilyView</h3><p>
Ensembl 
'<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=familyview">FamilyView</a>'
provides a list of closely related Ensembl gene predictions together
with a consensus family description and shows the chromosomal location
of family members on a karyotype ideogram. It also provides a list of
vertebrate <a href="http://www.ebi.ac.uk/uniprot/" target="external">UniProt</a> sequences and Ensembl protein predictions from other species that have been used to define the family. 
It therefore provides a way of exploring orthologues and closely related homologues across a range of animal species. 
</p>


<h5 id="EnsemblProteinFamilyMembers">Ensembl Protein Family Members</h5>

<ul>

<li>
<p>
The karyotype ideogram shows the chromosomal locations of the genes that produce proteins in the Ensembl family. 
Pointing to a red arrow pops up a menu with a link to Ensembl 
'<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=geneview">GeneView</a>' and 
'<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=contigview">ContigView</a>' for this gene. 
When jumping to 'ContigView' a green background in the 
'<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=contigview#DetailedView">Detailed View</a>' panel will mark the gene. 
</p>
</li>

<li>
<p>
If the Java runtime environment (<a href="http://java.sun.com/jre/" target="external">JRE</a>) is properly installed on your computer, then buttons will produce a new window with pre-calculated <b>multiple alignments</b> of family members displayed in 
<a href="http://www.jalview.org/" target="external">JalView</a>. 
You can choose to display just the Ensembl peptides (all species), or to display these plus all the 
<a href="http://www.ebi.ac.uk/uniprot/" target="external">UniProt</a>
members of the family. You can also export a text file with the
alignments of all the family members - a wide range of formats is
available from the pull-down menu. </p>
</li>

<li>
<p>
A table with links to each Ensembl gene in this family, with its gene description. 
Note that these are genes from the current species only. 
</p>
</li>

<li>
<p>
Tables with links to all 
<a href="http://www.ebi.ac.uk/swissprot/" target="external">UniProt/Swiss-Prot</a> and 
<a href="http://www.ebi.ac.uk/trembl/" target="external">UniProt/TrEMBL</a> entries that are part of the cluster. 
</p>
</li>

<li>
<p>
Tables with links to all <b>Ensembl peptides of other species</b>
that are part of the cluster. If an Ensembl species is not shown, no
family members have been identified from that species. Note that this
section shows peptides, not genes, and some entries may represent
alternative predicted products of the same gene. Navigate to a '<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=familyview">FamilyView</a>' page in the other species to check the number of different genes. 
</p>
</li>

</ul>

<p>
The link '<b>Export a list</b> of genes containing this family' will take you to the 
'<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=martview">EnsMart</a>' data mining system. 
The protein family will already have been used as a filter, and you will be ready to export the list of genes. 
Similarly, the link 'dump this family as <b>FASTA</b>' takes you to an EnsMart page ready to export the set of sequences. 
</p>


<h5 id="Clustering">Clustering</h5>

<p>
The protein family database is generated by running the Markov Clustering 
(<a href="http://micans.org/mcl/" target="external">MCL</a>) algorithm 
(<a href="#nb1" name="nh1">1</a>, <a href="#nb2" name="nh2">2</a>, <a href="#nb3" name="nh3">3</a>, <a href="#nb4" name="nh4">4</a>) as initially proposed by A.J. Enright, S. van Dongen and C.A. Ouzounis 
(<a href="#nb5" name="nh5">5</a>).
Prior to the clustering process, an all-against-all BLASTP sequence
similarity search is run on the super-set of all Ensembl protein
predictions of all species, together with all metazoan sequences from <a href="http://www.ebi.ac.uk/swissprot/" target="external">UniProt/Swiss-Prot</a> and 
<a href="http://www.ebi.ac.uk/trembl/" target="external">UniProt/TrEMBL</a>, to establish similarities. 
Using these similarities, protein family <b>clusters</b> are established running the 
<a href="http://micans.org/mcl/" target="external">MCL</a> algorithm. 
</p>

<p>
For each cluster thus obtained, a <b>Consensus Annotation</b> is automatically generated from the 
<a href="http://www.ebi.ac.uk/swissprot/" target="external">UniProt/Swiss-Prot</a> and 
<a href="http://www.ebi.ac.uk/trembl/" target="external">UniProt/TrEMBL</a> description lines of all UniProt members of each cluster using a longest common sub-string approach. 
</p>

<p>
The <b>annotation confidence score</b> is the percentage of UniProt family members that have this annotation, or part of it, in UniProt.
Note that only family members with 'informative' UniProt descriptions are taken into account. 
If the description covers less than 40% of these UniProt members, the family description is assigned 'AMBIGUOUS'. 
</p>


<h5 id="References">References</h5>

<p>
<a href="#nh1" name="nb1"><b>1</b></a>. 
Stijn van Dongen
<br>
<b>Graph Clustering by Flow Simulation.</b>
<br>
PhD thesis, University of Utrecht, May 2000
<br>
[<a href="http://www.library.uu.nl/digiarchief/dip/diss/1895620/inhoud.htm" target="external">Full text</a>]
</p>

<p>
<a href="#nh2" name="nb2"><b>2</b></a>. 
Stijn van Dongen
<br>
<b>A cluster algorithm for graphs.</b>
<br>
Technical Report INS-R0010, 
National Research Institute for Mathematics and Computer Science in the Netherlands, 
Amsterdam, 
May 2000
<br>
[<a href="http://www.cwi.nl/ftp/CWIreports/INS/INS-R0010.ps.Z" target="external">PostScript file</a>]
</p>

<p>
<a href="#nh3" name="nb3"><b>3</b></a>. 
Stijn van Dongen
<br>
<b>A stochastic uncoupling process for graphs.</b>
<br>
Technical Report INS-R0011, 
National Research Institute for Mathematics and Computer Science in the Netherlands, 
Amsterdam, 
May 2000
<br>
[<a href="http://www.cwi.nl/ftp/CWIreports/INS/INS-R0011.ps.Z" target="external">PostScript file</a>]
</p>

<p>
<a href="#nh4" name="nb4"><b>4</b></a>. 
Stijn van Dongen
<br>
<b>Performance criteria for graph clustering and Markov cluster experiments.</b>
<br>
Technical Report INS-R0012, 
National Research Institute for Mathematics and Computer Science in the Netherlands, 
Amsterdam, 
May 2000
<br>
[<a href="http://www.cwi.nl/ftp/CWIreports/INS/INS-R0012.ps.Z" target="external">PostScript file</a>]
</p>

<p>
<a href="#nh5" name="nb5"><b>5</b></a>. 
Anton J. Enright, 
Stijn van Dongen and 
Christos A. Ouzounis
<br>
<b>An efficient algorithm for large-scale detection of protein families.</b>
<br>
Nucleic Acids Res. 2002 Apr 1;30(7):1575-1584.
<br>
[<a href="http://nar.oupjournals.org/cgi/content/abstract/30/7/1575" target="external">Abstract</a>] 
[<a href="http://nar.oupjournals.org/cgi/content/full/30/7/1575" target="external">Full text</a>] 
</p>

<p>
The search box at the top of the page allows you to search for any identifier present in Ensembl. 
For detailed instructions see the Ensembl 
'<a href="http://www.ensembl.org/Canis_familiaris/helpview?se=1&amp;kw=textview">TextView</a>' page. 
</p>
<br><br>
  
</td></tr></tbody></table>

</body></html>