<html><head><title>Eh?: geneview</title>
    


<li id="OrthologuePrediction">
<b>Orthologue Prediction</b> - a link to the closest putative orthologue genes in other species. 
Orthologue prediction are computed on a paired species basis between all species released in Ensembl.<br>
For each orthologue prediction, a link is available that leads to a corresponding Ensembl 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=multicontigview">MultiContigView</a>' page, 
which allows display of both genome sequence and annotation information for several species simultaneously. A link to
the peptide alignment between the 2 orthologue genes is also available.
<p>

<p>
In our analysis, in case of alternative isoforms for a gene, we only consider the longest peptide isoform,
so that we have a 1-to-1 gene/peptide relation. The orthologous genes represent the peptide best reciprocal
BLAST hits for the two considered species (Best Reciprocal Hit: BRH) using a discriminative function based on
score, e-value, % identity and finally % positivity.
We can obtain Unique BRH (UBRH) when the discrimitive is unambigous and returns you only one BRH.
But sometimes it can return more than one, then you obtain multiple BRHs (MBRH, which stands for one
of (M)any BRHs).

For closely releated species, i.e. inside vertebrate phylum, or inside arthropoda phylum, 
where some gene order conservation is expected (or synteny), we identify additional orthologous
pairs obtained by a combination of reciprocal BLAST and location information. These additional
orthologue prediction are named Reciprocal Hit based on Synteny (RHS). </p>

<p>
Human/Chimpanzee exception:
The human/chimpanzee orthlogue prediction were obtained in a completely different manner.
The chimpanzee genome is low coverage 4X assembly of poor quality to generate a gene set on the classical
Ensembl gene build pipeline. The gene set produced by Ensembl has been generated by "projection" of human
genes to the chimpanzee genome through a whole genome BLASTz alignments between human and chimp filtered for
orthologue sequence alignments. So we have de facto the human/chimpanzee orthologue genes and are Derived 
from Whole Genome Alignments (DWGA).</p>

<p>
Clicking on the putative orthologue gene will take you to a 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=geneview">GeneView</a>' display of that gene within the web site for the other species. 
See also the 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=syntenyview">SyntenyView</a>' help page. 
Where no orthologue predictions are shown, you may wish to explore the Ensembl protein family (see below). 
</p>

<p>
dN (number of non-synonymous substitutions / number of non-synonymous sites) and 
dS (number of synonymous substitutions / number of synonymous sites) values were generated using the <b>codeml</b> program included in the 
<a href="http://abacus.gene.ucl.ac.uk/software/paml.html" target="external">PAML</a> package 
(<a href="#nb1" name="nh1">1</a>). 
With the parameters we have used, <b>codeml</b> performs pair wise Maximum Likelihood calculations of dN and dS for each orthologous pair. 
We have used the F3x4 codon evolution model 
(<a href="#nb2" name="nh2">2</a>).
This takes into account both the bias deriving from the different
probabilities of transition (T&lt;-&gt;C and A&lt;-&gt;G) versus
transversion (T/C&lt;-&gt;A/G) mutations, and the bias due to different
nucleotide frequencies at the three codon positions (for more details on the paramerters, the codeml parameter file can be found in ensembl-compara/scripts/homology/codeml.ctl). </p>

<p>
dN and dS values are only provided for orthologues from some species
pairs, i.e. human/mouse, human/rat, and mouse/rat. Orthologues for
other species pairs are too divergent for dS to be an accurate measure.
Most synonymous sites will have be subjected to more than one mutation
and ancestral changes cannot be reliably inferred from extant
sequences, (i.e. dS is saturated). </p><p>

</p><p>
Orthologue predictions for human/mouse, human/rat and mouse/rat may not
be perfect. Incorrect assignments will manifest anomalously high dS
values. We have, therefore, applied a cut-off of twice the median value
of all dS for each species pair as the criterion for displaying the
dN/dS ratio. Predicted orthology relationships with dS above this
threshold are likely to be errors. </p>

<p>
The following dS threshold values have been used: 
</p>

<pre>                   dS threshold

   human/mouse        1.25560
   human/rat          1.26930
   mouse/rat          0.40780
</pre>

<p>
Therefore, dN/dS ratios for human/mouse orthologues are only displayed when dS &lt;= 1.25560, otherwise "--". 
</p>

<h5 id="References">References</h5>

<p>
<a href="#nh1" name="nb1"><b>1</b></a>. 
Yang, Z. (1997)<br>
"PAML: a program package for phylogenetic analysis by maximum likehood."<br>
Comput. Appl. Biosci. 1997 13: 555-556.<br>
</p>

<p>
<a href="#nh2" name="nb2"><b>2</b></a>. 
Goldman, N. &amp; Zang, Y. (1994)<br>
"A codon-based model of nucleotide substitution for protein-coding DNA sequences."<br>
Mol. Biol. Evol. 11, 725-736.<br>
</p>

<p>
<a href="#nh3" name="nb3"><b>3</b></a>. 
Stein, LD et al. (2003)<br>
"The Genome Sequence of <i>Caenorhabditis briggsae</i>: A Platform for Comparative Genomics."
PLoS Biol. 1, 166-192.
</p>
</li>

<li id="ParaloguePrediction">
<b>Paralogue Prediction</b>
<p>The paralogous genes presently displayed within Ensembl are best defined as 'recently duplicated genes'.  For each organism where these recent duplications have been identified, this has be done so using a particular definition of what is recent.  For example, recent human duplications are those that appear to have occurred since the divergence of the human and rodent lineages.  Currently, these gene duplications are identified for just three species, human, mouse and rat.  This will be expanded to the full range of Ensembl species in the very near future.</p>
<p>The recent duplications for each species were detected using the combination of homology searches followed by genetic distance-based filtering.  For each organism, potential recent duplicates of each gene were derived using a simple BLAST homology search.  Where it made sense to do so, the database used for the homology search included not only all genes from the organism in question, but also genes from species that represent a useful outgroup.  Matches with outgroup sequences helped provide a dynamic cut-off during later filtering.  In the case of human genes, the closest match to a mouse or rat gene provided a cut-off point defining the time boundary for what constitutes a 'recent' duplication.  For genes that did not have a homologous sequence from an outgroup species, an arbritrary cutoff was applied.</p>
<p>Once a set of homologous genes was obtained by BLAST search, each hit
sequence was aligned with the query (a codon based alignment was
generated).  The Ks genetic distance was then calculated for each
query-hit gene pair using the Nei and Gojobori method (<a href="#ref1"
name="ref1">4</a>) as implemented
in the codeml program from the PAML software package (<a href="#ref2" name="ref2">5</a>).  Intra-species
hit sequences were considered to be recent duplicates if they were
more closely related than the nearest outgroup sequence (or, in the
case of no related inter-species genes, if they were more closely
related distance than an arbritrary genetic distance cut-off).</p>

<h5>Pipeline Steps - Per Gene:</h5>
<ul>
<li>Use BLAST to identify homologous genes both intra-species and with
certain outgroup species.</li>
<li>Perform pairwise codon-based alignment of query gene with each
homologous gene identified from BLAST search.</li>
<li>Calculate pairwise genetic distances between aligned query gene and
homologues.</li>
<li>Filter homologous genes according to genetic distance, keeping only
those genes from the same species as the query gene, with a smaller
genetic distance than the most related gene from an outgroup species
(or apply an artificial cutoff where outgroup matches aren't found).</li>
<li>Final list of gene matched with recently duplicated, paralogous genes.</li>
</ul>

<h5>Species specific outgroups and arbritrary cutoffs</h5>

<table border=1 cellspacing=0 cellpadding=1>
<tr><th>Species</th><th>Outgroup species</th><th>Arbritrary Cutoff</th></tr>
<tr valign=top><td><i>Homo sapiens</i></td><td><i>Mus musculus<br>Rattus Norvegicus</i></td><td align=center>0.6</td></tr>
<tr valign=top><td><i>Mus musculus</i></td><td><i>Homo sapiens</i></td><td align=center>0.6</td></tr>
<tr valign=top><td><i>Rattus norvegicus</i></td><td><i>Homo sapiens</i></td><td align=center>0.6</td></tr>
</table>

<h5>References</h5>
<p>
<a href="#ref1" name="ref1"><b>4</b></a>. 
Nei, M. &amp; Gojobori, T. (1986)<br>
"Simple methods for estimating the numbers of synonymous and
nonsynonymous nucleotide substitutions."<br>
Mol. Biol. Evol. 3:418-426<br>
</p>
<p>
<a href="#ref2" name="ref2"><b>5</b></a>. 
Yang, Z. (1997)<br>
"PAML: a program package for phylogenetic analysis by maximum likehood."<br>
Comput. Appl. Biosci. 1997 13: 555-556.<br>
</p>
</li>


</body></html>
